File: tools/xgmtool/src/ym2612.c
#include <stdlib.h>
#include <stdbool.h>

#include "../inc/vgmcom.h"
#include "../inc/ym2612.h"

#define DUALS_SIZE      7


const static int duals[DUALS_SIZE][2] =
{
    {0x24, 0x25}, {0xA4, 0xA0}, {0xA5, 0xA1}, {0xA6, 0xA2}, {0xAC, 0xA8}, {0xAD, 0xA9},
    {0xAE, 0xAA}
};


YM2612* YM2612_create()
{
    YM2612* result;

    result = malloc(sizeof(YM2612));

    YM2612_clear(result);

    return result;
}

YM2612* YM2612_copy(YM2612* source)
{
    YM2612* result;
    int i;

    result = YM2612_create();

    for (i = 0; i < 0x100; i++)
    {
        result->registers[0][i] = source->registers[0][i];
        result->registers[1][i] = source->registers[1][i];
        result->init[0][i] = source->init[0][i];
        result->init[1][i] = source->init[1][i];
    }

    return result;
}

void YM2612_clear(YM2612* source)
{
    int i;

    for (i = 0; i < 0x20; i++)
    {
        source->registers[0][i] = -1;
        source->registers[1][i] = -1;
        source->init[0][i] = false;
        source->init[1][i] = false;
    }
    for (i = 0x20; i < 0x100; i++)
    {
        source->registers[0][i] = -1;
        source->registers[1][i] = -1;
        source->init[0][i] = false;
        source->init[1][i] = false;
    }
}

void YM2612_initialize(YM2612* source)
{
    int i;

    for (i = 0; i < 0x20; i++)
    {
        source->registers[0][i] = 0;
        source->registers[1][i] = 0;
        source->init[0][i] = true;
        source->init[1][i] = true;
    }
    for (i = 0x20; i < 0x100; i++)
    {
        source->registers[0][i] = 0xFF;
        source->registers[1][i] = 0xFF;
        source->init[0][i] = true;
        source->init[1][i] = true;
    }
}

int YM2612_get(YM2612* source, int port, int reg)
{
    if (YM2612_canIgnore(port, reg))
        return 0;

    return source->registers[port][reg];
}

bool YM2612_set(YM2612* source, int port, int reg, int value)
{
    if (YM2612_canIgnore(port, reg))
        return false;

    int newValue = value;

    if (port == 0)
    {
        // special case of KEY ON/OFF register
        if (reg == 0x28)
        {
            int oldValue = source->registers[port][value & 7];
            newValue &= 0xF0;

            if (oldValue != newValue)
            {
                source->registers[port][value & 7] = newValue;
//                source->init[port][value & 7] = true;

                // always write when key state change
                return true;
            }

            return false;
        }

        // special case of Timer register --> ignore useless bits
        if (reg == 0x27)
            newValue &= 0xC0;
    }

    source->registers[port][reg] = newValue;
    source->init[port][reg] = true;

    return false;
}

bool YM2612_isSame(YM2612* source, YM2612* state, int port, int reg)
{
    if ((source->init[port][reg] == false) && (state->init[port][reg] == false))
        return true;

    return (source->init[port][reg] || YM2612_canIgnore(port, reg)) && (YM2612_get(source, port, reg) == YM2612_get(state, port, reg));
}

bool YM2612_isDiff(YM2612* source, YM2612* state, int port, int reg)
{
    return !YM2612_isSame(source, state, port, reg);
}


/**
 * Returns commands list to update to the specified YM2612 state
 */
List* YM2612_getDelta(YM2612* source, YM2612* state)
{
    List* result;
    int i, port, reg;

    result = createList();

    // do dual reg first
    for (i = 0; i < DUALS_SIZE; i++)
    {
        const int* dual = duals[i];
        const int reg0 = dual[0];
        const int reg1 = dual[1];

        // value is different
        if (YM2612_isDiff(source, state, 0, reg0) || YM2612_isDiff(source, state, 0, reg1))
        {
            // add commands
            addToList(result, VGMCommand_createYMCommand(0, reg0, YM2612_get(state, 0, reg0)));
            addToList(result, VGMCommand_createYMCommand(0, reg1, YM2612_get(state, 0, reg1)));
        }
        // port 1 too ?
        if (dual[0] > 0x30)
        {
            if (YM2612_isDiff(source, state, 1, reg0) || YM2612_isDiff(source, state, 1, reg1))
            {
                // add commands
                addToList(result, VGMCommand_createYMCommand(1, reg0, YM2612_get(state, 1, reg0)));
                addToList(result, VGMCommand_createYMCommand(1, reg1, YM2612_get(state, 1, reg1)));
            }
        }
    }

    for (port = 0; port < 2; port++)
    {
        for (reg = 0; reg < 0x100; reg++)
        {
            // can ignore or special case of KEY ON/OFF register
            if (YM2612_canIgnore(port, reg) || ((port == 0) && (reg == 0x28)))
                continue;

            // ignore dual reg
            if (YM2612_getDualReg(reg) != NULL)
                continue;

            // value is different --> add command
            if (YM2612_isDiff(source, state, port, reg))
                addToList(result, VGMCommand_createYMCommand(port, reg, YM2612_get(state, port, reg)));
        }
    }

    return result;
}

/**
 * Return true if write can be ignored
 */
bool YM2612_canIgnore(int port, int reg)
{
    switch (reg)
    {
        case 0x22:
        case 0x24:
        case 0x25:
        case 0x26:
        case 0x27:
        case 0x28:
        case 0x2B:
            return (port == 1);
    }

    if ((reg >= 0x30) && (reg < 0xB8))
        return ((reg & 3) == 3);

    return true;
}


/**
 * Return the dual entry for this reg (or NULL if no dual entry)
 */
int* YM2612_getDualReg(int reg)
{
    int i;

    for (i = 0; i < DUALS_SIZE; i++)
    {
        const int* dual = duals[i];

        if ((dual[0] == reg) || (dual[1] == reg))
            return (int*) dual;
    }

    return NULL;
}

File: tools/crates/xgmtool/src/ym2612.rs
use crate::vgm_command::VGMCommand;

const DUALS_SIZE: usize = 7;
const DUALS: [[u8; 2]; DUALS_SIZE] = [
    [0x24, 0x25], [0xA4, 0xA0], [0xA5, 0xA1], [0xA6, 0xA2],
    [0xAC, 0xA8], [0xAD, 0xA9], [0xAE, 0xAA],
];

/// YM2612 FM synthesizer state tracker
#[derive(Clone, Debug)]
pub struct YM2612 {
    pub registers: [[i32; 256]; 2],  // [port][reg]
    pub init: [[bool; 256]; 2],
}

impl YM2612 {
    pub fn new() -> Self {
        let mut ym = YM2612 {
            registers: [[-1; 256]; 2],
            init: [[false; 256]; 2],
        };
        ym.clear();
        ym
    }

    pub fn clear(&mut self) {
        for i in 0..256 {
            self.registers[0][i] = -1;
            self.registers[1][i] = -1;
            self.init[0][i] = false;
            self.init[1][i] = false;
        }
    }

    pub fn initialize(&mut self) {
        for i in 0..0x20 {
            self.registers[0][i] = 0;
            self.registers[1][i] = 0;
            self.init[0][i] = true;
            self.init[1][i] = true;
        }
        for i in 0x20..256 {
            self.registers[0][i] = 0xFF;
            self.registers[1][i] = 0xFF;
            self.init[0][i] = true;
            self.init[1][i] = true;
        }
    }

    pub fn get(&self, port: usize, reg: usize) -> i32 {
        if Self::can_ignore(port as i32, reg as i32) { return 0; }
        self.registers[port][reg]
    }

    /// Set register value. Returns true if this is a KEY write that changed state.
    pub fn set(&mut self, port: usize, reg: usize, value: i32) -> bool {
        if Self::can_ignore(port as i32, reg as i32) { return false; }

        let mut new_value = value;

        if port == 0 {
            // KEY ON/OFF register
            if reg == 0x28 {
                let key_reg = (value & 7) as usize;
                let old_value = self.registers[0][key_reg];
                new_value = value & 0xF0;
                if old_value != new_value {
                    self.registers[0][key_reg] = new_value;
                    return true;
                }
                return false;
            }
            // Timer register
            if reg == 0x27 {
                new_value = value & 0xC0;
            }
        }

        self.registers[port][reg] = new_value;
        self.init[port][reg] = true;
        false
    }

    pub fn is_same(&self, other: &YM2612, port: usize, reg: usize) -> bool {
        if !self.init[port][reg] && !other.init[port][reg] { return true; }
        (self.init[port][reg] || Self::can_ignore(port as i32, reg as i32))
            && self.get(port, reg) == other.get(port, reg)
    }

    pub fn is_diff(&self, other: &YM2612, port: usize, reg: usize) -> bool {
        !self.is_same(other, port, reg)
    }

    /// Generate VGM commands to transition from self state to `target` state
    pub fn get_delta(&self, target: &YM2612) -> Vec<VGMCommand> {
        let mut result = Vec::new();

        // Dual registers first
        for dual in &DUALS {
            let reg0 = dual[0] as usize;
            let reg1 = dual[1] as usize;

            if self.is_diff(target, 0, reg0) || self.is_diff(target, 0, reg1) {
                result.push(VGMCommand::create_ym_command(0, dual[0], target.get(0, reg0) as u8));
                result.push(VGMCommand::create_ym_command(0, dual[1], target.get(0, reg1) as u8));
            }
            if dual[0] > 0x30 {
                if self.is_diff(target, 1, reg0) || self.is_diff(target, 1, reg1) {
                    result.push(VGMCommand::create_ym_command(1, dual[0], target.get(1, reg0) as u8));
                    result.push(VGMCommand::create_ym_command(1, dual[1], target.get(1, reg1) as u8));
                }
            }
        }

        // Other registers
        for port in 0..2 {
            for reg in 0..256 {
                if Self::can_ignore(port as i32, reg as i32) { continue; }
                if port == 0 && reg == 0x28 { continue; } // KEY register
                if Self::get_dual_reg(reg as u8).is_some() { continue; }

                if self.is_diff(target, port, reg) {
                    result.push(VGMCommand::create_ym_command(
                        port as i32, reg as u8, target.get(port, reg) as u8,
                    ));
                }
            }
        }

        result
    }

    /// Returns true if writes to this register can be ignored
    pub fn can_ignore(port: i32, reg: i32) -> bool {
        match reg {
            0x22 | 0x24 | 0x25 | 0x26 | 0x27 | 0x28 | 0x2B => port == 1,
            _ => {
                if reg >= 0x30 && reg < 0xB8 {
                    (reg & 3) == 3
                } else {
                    true
                }
            }
        }
    }

    /// Return the dual register pair for this reg, if any
    pub fn get_dual_reg(reg: u8) -> Option<[u8; 2]> {
        for dual in &DUALS {
            if dual[0] == reg || dual[1] == reg {
                return Some(*dual);
            }
        }
        None
    }
}
